dali/kernels/tensor_view.h

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_TENSOR_VIEW_H_
+#define DALI_KERNELS_TENSOR_VIEW_H_
+
+#include "shape.h"
+
+namespace tensor {
+
+struct EmptyBackendTag {};
+
+template <typename Backend, typename DataType, int ndim>
+struct TensorView;
+
+template <typename Backend, typename DataType>
+struct TensorView<Backend, DataType, DynamicDimensions> {
+  TensorView() : data(nullptr), shape() {}
+
+  template <int ndim>
+  TensorView(DataType *data, TensorShape<ndim> shape) : data(data), shape(shape) {}
+  TensorView(const TensorView &) = default;
+  TensorView(TensorView &&) = default;
+  TensorView &operator=(const TensorView &) = default;
+  TensorView &operator=(TensorView &&) = default;
```

Comment:
Default move for dynamic TensorView (and any TensorListView) leaves me with slight distaste. We move out the shape information, but keep the original data pointer. Maybe we should nullify the RHS for completeness? In most cases the compiler will elide the assignment anyway, since all relevant functions are inline and it will be able to detect dead store.

Reply:
Done